apps/s_time.c

```diff
@@ -240,16 +224,10 @@ int s_time_main(int argc, char **argv)
                                    www_path);
             if (buf_len <= 0 || SSL_write(scon, buf, buf_len) <= 0)
                 goto end;
-            while ((i = SSL_read(scon, buf, sizeof(buf))) > 0 ||
-                        SSL_get_error(scon, i) == SSL_ERROR_WANT_READ ||
-                        SSL_get_error(scon, i) == SSL_ERROR_WANT_WRITE)
```

Comment:
> ./openssl s_server -www
vs.
./openssl s_time -new
=>
5134 connections in 8.30s; 618.55 connections/user sec, bytes read 0
5134 connections in 31 real seconds, 0 bytes read per connection
./openssl s_time -reuse
=> (as expected SSL_session_reused() = 0)
5049 connections in 8.36s; 603.95 connections/user sec, bytes read 0
5049 connections in 31 real seconds, 0 bytes read per connection

These numbers are as expected, i.e. we get less connections/sec with the attempted reuse than without. In both cases we are doing a full handshake (we're not actually reusing)...but we've gone through some of the hoops of attempting to reuse. At least that's my guess.


> ./openssl s_time -www /test -new
=>
4435 connections in 8.31s; 533.69 connections/user sec, bytes read 21953250
4435 connections in 31 real seconds, 4950 bytes read per connection

Also expected. Less connections/sec because now we are actually transferring data.

> Note: the connections/user sec are almost identical, does that mean the client
has no less work to do for a reuse, but the server ?

I'm not sure what you mean. So far in the above numbers no reuse has actually happened (SSL_session_reused() == 0)


> ./openssl s_time -www /test -reuse
=> (as expected SSL_session_reused() = 1)
8734 connections in 14.90s; 586.17 connections/user sec, bytes read 38840098
8734 connections in 31 real seconds, 4447 bytes read per connection

For the first time connections/sec goes up because we are now really doing reuse, so this is as expected.

> now I start the server with tls1_2
./openssl s_server -www -tls1_2

> ./openssl s_time -new
=>
4626 connections in 7.92s; 584.09 connections/user sec, bytes read 0
4626 connections in 31 real seconds, 0 bytes read per connection

> ./openssl s_time -reuse
=>
40253 connections in 16.34s; 2463.46 connections/user sec, bytes read 0
40253 connections in 31 real seconds, 0 bytes read per connection
Note: This is the only data point which is how I would expect it.

> ./openssl s_time -www /test -new
=>
4527 connections in 7.85s; 576.69 connections/user sec, bytes read 22413177
4527 connections in 31 real seconds, 4951 bytes read per connection

All of the above are largely as expected. Session reuse is faster than no session reuse. TLSv1.3 without session reuse is out performing TLSv1.2 without session reuse.

Slightly surprising is that TLSv1.2 www without session reuse is marginally better than the TLSv1.3 equivalent.

> ./openssl s_time -www /test -reuse
=> (yes SSL_session_reused() = 1 here)
725 connections in 0.87s; 833.33 connections/user sec, bytes read 2657125
725 connections in 31 real seconds, 3665 bytes read per connection

This one I cannot explain. TLSv1.2 www with session reuse is much worse than without???